### PR TITLE
Various fixes for database revision specifiers

### PR DESCRIPTION
**Pull request by fulghum**
Changing revision databases to be instantiated on-demand, instead of trying to track them in DoltDatabaseProvider. This simplifies how we manage revision databases and avoids revision dbs hanging around in memory after customers are done using them. The biggest UI changes for customers is that `show databases` will now no longer show all revision dbs that have ever been used, but instead will only include the single revision database a customer is currently using. 

Fixes: https://github.com/dolthub/dolt/issues/4067
Fixes: https://github.com/dolthub/dolt/issues/4031

### CODE REVIEW

File: integration-tests/bats/deleted-branches.bats [hunk 1]
```diff
@@ -73,34 +73,6 @@ make_it() {
     server_query "dolt_repo_$$/main" 1 "SELECT * FROM test" "id\n"
 }
 
-@test "deleted-branches: can DOLT_CHECKOUT on SQL connection with existing branch revision specifier when dolt_default_branch is invalid" {
```
Comment (fulghum) [inline integration-tests/bats/deleted-branches.bats:76]: These two tests happened to work, because there was an existing session that `use`d the branch-qualified database revision, before the default branch was corrupted. Since this change stopped tracking database revisions as databases in the DatabaseProvider, these tests started failing. 

I spent some time digging through this, but ended up deciding this was not a super important edge case to continue supporting, so I removed the tests. I'm happy to give it some more time if people think this is important though.

Comment (zachmu) [reply integration-tests/bats/deleted-branches.bats:76]: I think rather than deleting, the right approach is just to lower the bar for this test so the behavior is still defined and we know it doesn't panic. Basically anything can happen, we just want to make sure it's not a fatal error.

Comment (zachmu) [reply integration-tests/bats/deleted-branches.bats:76]: Same for the one below obv

File: integration-tests/bats/sql-server.bats [hunk 2]
```diff
@@ -1258,7 +1258,7 @@ END""")
     # can't drop a branch-qualified database name
     run server_query "" 1 "drop database \`test2/newbranch\`"
     [ "$status" -ne 0 ]
-    [[ "$output" =~ "database not found: test2/newbranch" ]] || false
+    [[ "$output" =~ "unable to drop revision database: test2/newbranch" ]] || false
 
     server_query "" 1 "drop database TEST2"
```
Comment (zachmu) [inline integration-tests/bats/sql-server.bats:1263]: Would be cool to have some bats tests showing the output for `show databases` after something like `use mydb/HEAD~`

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go [hunk 3]
```diff
@@ -125,6 +125,146 @@ var DescribeTableAsOfScriptTest = queries.ScriptTest{
 	},
 }
 
+var DoltRevisionDbScripts = []queries.ScriptTest{
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries.go:128]: Any tests for things like `HEAD~`? What about commit hashes?

**Review by zachmu — APPROVED** [review]
Overall looks really good, just some suggestions for tests.

Did `use mydb/HEAD~` not make it into this?

File: go/libraries/doltcore/sqle/enginetest/dolt_queries.go (hunk 3 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/enginetest/dolt_queries.go:128]: No support for `HEAD~` refs yet, but I'd be more than happy to add that in a separate PR. Seems like a very natural and useful extension. 

I made BATS tests covering commit hashes in db rev specs, but didn't come up with a good way to get the commit hash and easily use it in the enginetests with the `ScriptTest` structures. I figured the BATS coverage was decent, but I'll try adding a non-ScriptTest and see if I can get some enginetest coverage for commits, too.
